src/platform_impl/android/mod.rs

```diff
@@ -491,163 +504,181 @@ impl<T: 'static> EventLoop<T> {
-            }
+        // Consider the possibility that the `StartCause::Init` iteration could
+        // request to Exit
+        if !matches!(self.control_flow, ControlFlow::ExitWithCode(_)) {
+            self.poll_events_with_timeout(timeout, &mut callback);
+        }
+        if let ControlFlow::ExitWithCode(code) = self.control_flow {
+            self.loop_running = false;
+
+            let mut dummy = self.control_flow;
+            sticky_exit_callback(
+                event::Event::LoopDestroyed,
+                self.window_target(),
+                &mut dummy,
+                &mut callback,
+            );
+
+            PumpStatus::Exit(code)
+        } else {
+            PumpStatus::Continue
+        }
+    }
 
-            let app = self.android_app.clone(); // Don't borrow self as part of poll expression
-            app.poll_events(timeout, |poll_event| {
-                let mut main_event = None;
-
-                match poll_event {
-                    android_activity::PollEvent::Wake => {
-                        // In the X11 backend it's noted that too many false-positive wake ups
-                        // would cause the event loop to run continuously. They handle this by re-checking
-                        // for pending events (assuming they cover all valid reasons for a wake up).
-                        //
-                        // For now, user_events and redraw_requests are the only reasons to expect
-                        // a wake up here so we can ignore the wake up if there are no events/requests.
-                        // We also ignore wake ups while suspended.
-                        pending_redraw |= self.redraw_flag.get_and_reset();
-                        if !self.running
-                            || (!pending_redraw && !self.user_events_receiver.has_incoming())
-                        {
-                            return;
+    fn poll_events_with_timeout<F>(&mut self, mut timeout: Option<Duration>, mut callback: F)
+    where
+        F: FnMut(event::Event<'_, T>, &RootELW<T>, &mut ControlFlow),
+    {
+        let start = Instant::now();
+
+        self.pending_redraw |= self.redraw_flag.get_and_reset();
+
+        timeout =
+            if self.running && (self.pending_redraw || self.user_events_receiver.has_incoming()) {
+                // If we already have work to do then we don't want to block on the next poll
+                Some(Duration::from_millis(0))
+            } else {
+                let control_flow_timeout = match self.control_flow {
+                    ControlFlow::Wait => None,
+                    ControlFlow::Poll => Some(Duration::from_millis(0)),
+                    ControlFlow::WaitUntil(wait_deadline) => {
+                        if wait_deadline > start {
+                            Some(wait_deadline - start)
+                        } else {
+                            Some(Duration::from_millis(0))
                         }
```

Comment:
This should be enough, the `sub` saturates.

```suggestion
                        Some(wait_deadline - start)
```

Reply:
hmmm, that seems really non-obvious to me. 

Looking at the implementation it actually looks like the behaviour has changed in the past and it even warns that it could change again:

```rust
    /// Returns the amount of time elapsed from another instant to this one,
    /// or zero duration if that instant is later than this one.
    ///
    /// # Panics
    ///
    /// Previous rust versions panicked when `other` was later than `self`. Currently this
    /// method saturates. Future versions may reintroduce the panic in some circumstances.
    /// See [Monotonicity].
    ///
    /// [Monotonicity]: Instant#monotonicity
    fn sub(self, other: Instant) -> Duration {
        self.duration_since(other)
    }
```

I'll just change this to use `ZERO` I think since I missed that change in the android backend.

Reply:
Well, you can straight call `saturating_sub` instead of `-` and `if`.

Reply:
It doesn't look like there's a saturating_sub for `Instances`

There's [saturating_duration_since](https://doc.rust-lang.org/std/time/struct.Instant.html#method.saturating_duration_since) and I guess we could do `wait_deadline.saturating_duration_since(start)` but that also feels like a back-to-front way of wording it that doesn't feel super clear.

Reply:
updated it to use `Some(wait_deadline.saturating_duration_since(start)),`